src/AstroDates.jl

```diff
@@ -216,113 +175,235 @@ const UNIX_EPOCH = Date(1970, 1, 1)
 struct Time{T}
     hour::Int
     minute::Int
-    second::T
+    second::Int
+    fraction::T
 
-    function Time(hour, minute, second::T) where T
+    function Time(hour, minute, second, fraction::T) where {T}
         if hour < 0 || hour > 23
             throw(ArgumentError("`hour` must be an integer between 0 and 23."))
         elseif minute < 0 || minute > 59
             throw(ArgumentError("`minute` must be an integer between 0 and 59."))
-        elseif second < 0 || second >= 61.0
-            throw(ArgumentError("`second` must be a float between 0 and 61."))
+        elseif second < 0 || second >= 61
+            throw(ArgumentError("`second` must be an integer between 0 and 61."))
+        elseif fraction < 0 || fraction > 1
+            throw(ArgumentError("`fraction` must be a number between 0 and 1."))
         end
 
-        return new{T}(hour, minute, second)
+        return new{T}(hour, minute, second, fraction)
     end
 end
 
-function Time(second_in_day_a, second_in_day_b)
-    carry = floor(Int, second_in_day_b)
-    wholeseconds = second_in_day_a + carry
-    fractional = second_in_day_b - carry
+function Time(hour, minute, second)
+    sec, frac = divrem(second, 1)
+    return Time(hour, minute, sec, frac)
+end
 
-    # range check
-    if wholeseconds < 0 || wholeseconds > 86400
+function Time(secondinday, fraction)
+    if secondinday < 0 || secondinday > 86400
         throw(ArgumentError("Seconds are out of range"))
     end
 
-    # extract the time components
-    hour = wholeseconds ÷ 3600
-    wholeseconds -= 3600 * hour
-    minute = wholeseconds ÷ 60
-    wholeseconds -= 60 * minute
-    second = wholeseconds + fractional
+    hour = secondinday ÷ 3600
+    secondinday -= 3600 * hour
+    minute = secondinday ÷ 60
+    secondinday -= 60 * minute
 
-    return Time(hour, minute, second)
+    return Time(hour, minute, secondinday, fraction)
 end
 
-Time(t::Dates.Time) = Time(Dates.hour(t), Dates.minute(t),
-                           Dates.second(t) + 1e-3Dates.millisecond(t))
-Dates.Time(t::Time) = Dates.Time(hour(t), minute(t), second(t), millisecond(t))
+function Base.isapprox(a::Time, b::Time; kwargs...)
+    return a.hour == b.hour &&
+           a.minute == b.minute &&
+           isapprox(a.fraction + a.second, b.fraction + b.second; kwargs...)
+end
 
-const H00 = Time(0, 0, 0.0)
-const H12 = Time(12, 0, 0.0)
+function subsecond(fraction, n, r)
```

Comment:
This stumped me for a while. See also the method below. As a result, we now get sensible values for milli-/micro-/nanoseconds as if they were integral values and hiding the inherent floating point inaccuracies.

```julia
julia> subsecond(0.1839999999999975, 9)
0

julia> subsecond(0.1839999999999975, 3)
184

julia> subsecond(0.123456789, 9)
789

julia> subsecond(0.123456789, 3)
123
```